### PR TITLE
Catch exceptions in hooks defined by users

### LINKED ISSUE #1551: Exception handling for `on_worker_boot`
When an exception is raised in `on_worker_boot`, it will retry ad infinitum. I suppose there are situations when this is a good thing. E.g., a db instance is rebooting and a connection can't be made for a few seconds.

What's the best way to handle exceptions that should halt the server? How _do_ you actually halt the server in this case?

```
on_worker_boot do
  raise 'error'
end
```

I'm sorry if this is not the place for this question. I couldn't find a general discussion board.

### PR DESCRIPTION
**Pull request by andrykonchin**
### Description

Suppress exceptions in user defined hooks (like `on_worker_boot`). We have just to rescue and log them.

Example of logged message:

```
WARNING hook on_restart failed with exception (RuntimeError) Error from hook
% Traceback (most recent call last):
	14: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/parallel.rb:33:in `block (2 levels) in start'
	13: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest.rb:1026:in `run_one_method'
	12: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:93:in `run'
	11: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:211:in `with_info_handler'
	10: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest.rb:365:in `on_signal'
	 9: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:94:in `block in run'
	 8: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest.rb:270:in `time_it'
	 7: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:95:in `block (2 levels) in run'
	 6: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:195:in `capture_exceptions'
	 5: from /Users/andrykonchin/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/minitest-5.14.0/lib/minitest/test.rb:98:in `block (3 levels) in run'
	 4: from /Users/andrykonchin/projects/puma/test/test_config.rb:226:in `test_run_hooks_and_exception'
	 3: from /Users/andrykonchin/projects/puma/lib/puma/configuration.rb:279:in `run_hooks'
	 2: from /Users/andrykonchin/projects/puma/lib/puma/configuration.rb:279:in `each'
	 1: from /Users/andrykonchin/projects/puma/lib/puma/configuration.rb:281:in `block in run_hooks'
/Users/andrykonchin/projects/puma/test/test_config.rb:220:in `block (2 levels) in test_run_hooks_and_exception': Error from hook (RuntimeError)
```

Closes #1551

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by nateberkopec** [comment]
Makes sense and LGTM.

**Comment by andrykonchin** [comment]
There is only one issue with logging. I looked through the project and found out there are two options
* to directly write into `STDERR`
* to use `Events` (which could write either to `STDOUT`/`STDERR` or to strings).

I prefer the second one because it's easier to check in tests and `Events` provides more flexible interface - we can log error message as `log` message and optionally print exception stacktrace as `debug` message.

Please let me know if you prefer the first option.

File: History.md [hunk 1]
```diff
@@ -17,6 +17,7 @@
   * Preserve `BUNDLE_GEMFILE` env var when using `prune_bundler` (#1893)
   * Send 408 request timeout even when queue requests is disabled (#2119)
   * Rescue IO::WaitReadable instead of EAGAIN for blocking read (#2121)
+  * Rescue exceptions in hooks defined by users (on_worker_boot, after_worker_fork etc) (#1551)
```
Comment (nateberkopec) [inline History.md]: Let's say "Rescue _and log_ exceptions"

File: test/test_config.rb [hunk 2]
```diff
@@ -197,6 +197,112 @@ def test_config_raise_exception_on_sigterm
...
+      end
+    end
+    conf.load
+
+    conf.run_hooks :after_worker_fork, 'ARG', Puma::Events.strings
+    assert_equal messages, ['after_worker_fork is called with ARG']
+  end
+
+  def test_run_hooks_before_worker_boot
+    require 'puma/events'
```
Comment (nateberkopec) [inline test/test_config.rb]: Go ahead and require this at the top of the test file instead.

File: test/test_config.rb [hunk 3]
```diff
@@ -197,6 +197,112 @@ def test_config_raise_exception_on_sigterm
...
+    messages = []
+    conf = Puma::Configuration.new do |c|
+      c.before_fork do |a|
+        messages << "before_fork is called with #{a}"
+      end
+    end
+    conf.load
+
+    conf.run_hooks :before_fork, 'ARG', Puma::Events.strings
+    assert_equal messages, ['before_fork is called with ARG']
```
Comment (nateberkopec) [inline test/test_config.rb]: I'm really happy you added test coverage here. Can we abstract these tests a bit though?

e.g. 

```
assert_runs_hooks :before_fork 

def assert_runs_hooks(hook)
messages = []
    conf = Puma::Configuration.new do |c|
      c.send(hook) do |a|
        messages << "#{hook} is called with #{a}"
      end
    end
    conf.load

    conf.run_hooks hook, 'ARG', Puma::Events.strings
    assert_equal messages, ['#{hook} is called with ARG']
end
```

**Comment by nateberkopec** [comment]
Thanks for your contribution @andrykonchin!

**Pull request merged**